Bluepill-runner/Bluepill-runner/BPPacker.m

```diff
@@ -137,21 +145,67 @@ @implementation BPPacker
                     [BPUtils printInfo:INFO withString:@"Estimate not available for %@", test];
                 }
             }];
-            estimatedTimesByTestFilePath[xctFile.testBundlePath] = [NSNumber numberWithDouble:testBundleExecutionTime];
+            testEstimatedTimesByTestFilePath[xctFile.testBundlePath] = [NSNumber numberWithDouble:testBundleExecutionTime];
             totalTime += testBundleExecutionTime;
         }
     }
-    assert([estimatedTimesByTestFilePath count] == [xcTestFiles count]);
+    assert([testEstimatedTimesByTestFilePath count] == [xcTestFiles count]);
 
-    NSMutableArray<BPXCTestFile *> *bundles = [[NSMutableArray alloc] init];
+    [BPUtils printInfo:INFO withString:@"Stats before splitting..."];
+    NSMutableDictionary *inputDic = [NSMutableDictionary dictionary];
+    for(id key in testEstimatedTimesByTestFilePath) {
+        inputDic[[key substringFromIndex:[(NSString *)key rangeOfString:@"/" options:NSBackwardsSearch].location+1]] = [testEstimatedTimesByTestFilePath objectForKey:key];
+        NSLog(@"%@ = %@", [key substringFromIndex:[(NSString *)key rangeOfString:@"/" options:NSBackwardsSearch].location+1], [testEstimatedTimesByTestFilePath objectForKey:key]);
+    }
+
+    int minimumBundleTime = 180;
+    NSUInteger maxBundleTime = MAX(minimumBundleTime, totalTime / [config.numSims integerValue]);
+    [BPUtils printInfo:INFO withString:@"Max Bundle Time is around %lu seconds.", maxBundleTime];
+
+    // TODO: First of all check if the bundles need to be split or not (Hint: Based on numSims, current test bundle count and standard deviation of their execution times)
+    NSMutableArray<BPXCTestFile *> *testBundles = [[NSMutableArray alloc] init];
     for (BPXCTestFile *xctFile in xcTestFiles) {
-        BPXCTestFile *bundle = [xctFile copy];
-        bundle.skipTestIdentifiers = config.testCasesToSkip;
-        [bundle setEstimatedExecutionTime:estimatedTimesByTestFilePath[xctFile.testBundlePath]];
-        [bundles addObject:bundle];
+        NSArray *bundleTestsToRun = [[testsToRunByTestFilePath[xctFile.testBundlePath] allObjects] sortedArrayUsingSelector:@selector(compare:)];
+        double bundleEstimate = [testEstimatedTimesByTestFilePath[xctFile.testBundlePath] doubleValue];
+        if (bundleEstimate > maxBundleTime) {
+            NSUInteger packed = 0;
+            double splitExecTime = 0.0;
+            double sumOfSplits = 0.0;
+            for (int i = 0; i < [bundleTestsToRun count];) {
+                NSString *test = [bundleTestsToRun objectAtIndex:i];
+                NSMutableArray *myArray = [NSMutableArray array];
+                if ([testTimes objectForKey:test]) {
+                    splitExecTime += [testTimes[test] doubleValue];
+                    [myArray addObject:[NSNumber numberWithDouble:[testTimes[test] doubleValue]]];
+                }
+                i++;
+                if (splitExecTime > maxBundleTime || i >= [bundleTestsToRun count]) {
```

Comment:
So `i` is now in the next iteration? This pattern seems weird to me. Why can't we put the below part to the beginning of each iteration?